### DISCUSSION TITLE
img.Save err=png2vips: unable to read source source\n

### DISCUSSION THREAD

**Discussion by lifang218c** [libvips/libvips#2375] [Q&A]
## Questions, enhancements, tips, etc.

Please use libvips discussions for other topics.

https://github.com/libvips/libvips/discussions

## Bug report

**Describe the bug**
sometimes, it report this error, this is mycode:
err = C.vips_pngsave_go(img.VipsImage, &ptr, &imgsize, vipsConf.PngInterlaced, vipsConf.PngQuantize, vipsConf.PngQuantizationColors)

bellow is the error:
 img.Save err=png2vips: unable to read source source\n

**Environment**
(please complete the following information)
 - OS: debian, docker

**Comment by lifang218c** [comment]
it some times report this error, i don't know how to handle it. i set     
hope the writer can see. and give me some suggestions, thanks very much. my english is so bad.. sorry~

	C.vips_concurrency_set(1)
        C.vips_cache_set_max_mem(0)
	C.vips_cache_set_max(0)

**Comment by jcupitt** [comment]
Hi @lifang218c,

This can happen if the file is deleted between open and process. Here's what is happening:

1. Your code calls `new_from_file("x.png")`
2. libvips will read the file header, set the image metadata, close the file again, and return a `VipsImage` object.
3. Some time later, you call an operation which uses the `VipsImage` object, for example `pngsave`.
4. libvips will see that the `VipsImage` refers to a file on disk, and will reopen the file and start streaming pixels  from it as required.

If someone deletes or renames `"x.png"` between steps 2 and 4, you'll get this error. libvips does this close and reopen to try to minimise the number of open file descriptors.

**Comment by lifang218c** [comment]
thanks very much for your reply so quickly!

I am directly operating in the memory of the VipsImage, I have been down from the remote picture, and then stored in []byte, I need to watermark this image, and then call the img.save to get the resources of the picture.

The above four steps will not appear. I have no operation to rename or delete local image resources. Because I directly download remote image resources to memory for operation

if C.vips_pngsave_go(img.VipsImage, &ptr, &imgsize, vipsConf.PngInterlaced, vipsConf.PngQuantize, vipsConf.PngQuantizationColors) this happened, it retunrs the remote picture resources without watermark. how can i fix it? thanks. at the same time, i find that This error is a probability happend。 Probably 10%

**Comment by lifang218c** [comment]
Hi, @jcupitt , i find that this error is not always happend with a picture. 

sometimes, i get the right result. 
but sometimes, it report the error: img.Save err=png2vips: unable to read source source\n

i work with golang，i write this in a function:

`runtime.LockOSThread()
defer runtime.UnlockOSThread()
defer vipsCleanup()`
`func vipsCleanup() {
	C.vips_cleanup()
}`

I'm not sure if it's the impact， hope  your reply ， thanks

**Comment by jcupitt** [comment]
Hi again,

Then I think it must be a bug in the libvips binding you are using. The sequence of events might be:

1. Your code calls `new_from_memory(some_pointer)`
2. libvips will read the image header, set the metadata, and return a `VipsImage` object.
3. The Go GC reuses the area pointed to be `some_pointer`.
4. Some time later, you call an operation which uses the `VipsImage` object, for example `pngsave`.
5. libvips will now try to decompress pixels from `some_memory`, but there is no longer a valid PNG image at that address, and decode fails.

Your libvips binding needs to keep a reference to `some_memory` until you have finished using the `VipsImage` object. Which one are you using?

**Reply by lifang218c** [reply]
hi @jcupitt 
Your libvips binding needs to keep a reference to some_memory until you have finished using the VipsImage object. Which one are you using?

how can i fix this bug? 
`C.vips_cache_set_max_mem(0)
C.vips_cache_set_max(0)`

can i fix by set non-zero to fix this bug?

**Reply by lifang218c** [reply]
maybe you saied is right.  but i think the step 3 maybe not happened.
The Go GC will not gc the pointer. it untill this memory not be used.

**Comment by jcupitt** [comment]
It's very likely this is a bug in the libvips Go binding you are using, or perhaps in your code. I think you need to ask the author of the Go binding (I have never used Go).

I made you a simple C example:

```C
/* Compile with: 
 *      gcc -g -Wall buffer.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>

int
main( int argc, char **argv )
{
        gchar *buf;
        gsize len;
        int i;

        if( VIPS_INIT( argv[0] ) )
                vips_error_exit( NULL );

        if( !g_file_get_contents( argv[1], &buf, &len, NULL ) )
                vips_error_exit( NULL );

        for( i = 0; i < 100000; i++ ) {
                VipsImage *image;
                void *new_buf;
                size_t new_len;

                printf( "loop %d ...\n", i );

                if( !(image = vips_image_new_from_buffer( buf, len, "",
                        "access", VIPS_ACCESS_SEQUENTIAL,
                        NULL )) ) 
                        vips_error_exit( NULL );

                if( vips_pngsave_buffer( image, &new_buf, &new_len, 
                        "compression", 0,
                        NULL ) ) 
                        vips_error_exit( NULL );

                g_object_unref( image );
                g_free( new_buf );
        }

        g_free( buf );

        return( 0 );
}
```

I let this run for 3,000 iterations. There were no errors, and memory use remained stable at 350mb. You don't need to change the libvips cache. I don't think this is a libvips issue.

**Comment by lifang218c** [comment]
hi @jcupitt 

vips_error_exit. this function is set to the error not exists? i need todo this in my code? 

bellow is my code:

```go
       func ApplyWaterMark(ctx context.Context, imgDownload []byte, uid int, picId string, imgTempPath, serverWater string) ([]byte, context.CancelFunc) {
	childCtx, childSpan := tracex.CreateChild(ctx, "handlepic.ApplyWaterMark")
	if childSpan != nil {
		defer childSpan.End()
	}
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	defer vipsCleanup()

	img := new(vipsImage)
	defer img.Clear()
	imgdata, err := fileImageData(imgDownload)
	if err != nil {
		logx.Error("ApplyWaterMark fileImageData err=", err, " picId=", picId)
		return imgDownload, nil
	}

	if err := img.Load(imgdata.Data, imgdata.Type, 1, 1.0, 1); err != nil {
		logx.Error("ApplyWaterMark img.Load err=", err, " picId=", picId)
		return imgdata.Data, nil
	}

	if err = img.RgbColourspace(); err != nil {
		logx.Error("ApplyWaterMark img.RgbColourspace err=", err, " picId=", picId)
		return imgdata.Data, nil
	}

	width := img.Width()
	height := img.Height()
	if width >= 400 && width != 2 * height {
		if uid == 0 {
			logger.Info("ApplyWaterMark need uid")
			return imgdata.Data, nil
		}
		// 根据uid获取昵称，作为水印图，图片加水印
		userInfo, err := getUserInfo(childCtx, uid)
		if err != nil || userInfo.Nick == "" {
			userInfo = DefaultUserInfo
		}

		if userInfo.IsWatermarkingServer == 1 || serverWater == "must_have_water" {
			waterMarkSource, err := GenWaterMarkDataByNick(childCtx, userInfo.Nick, imgTempPath, uid, float64(width/37))
			if err != nil {
				return imgdata.Data, nil
			}

			defer DeleteUselessPic(waterMarkSource) // 清理签名图片的操作
			waterMarkSourceBuff , _ := ioutil.ReadFile(waterMarkSource)
			wm := new(vipsImage)
			defer wm.Clear()
			wmData := getImageData(waterMarkSourceBuff, imageTypePNG)
			if err := wm.Load(wmData.Data, wmData.Type, 1, 1.0, 1); err != nil {
				logx.Error("ApplyWaterMark wm.Load err=", err, " waterMarkSource=", waterMarkSource, " picId=", picId)
				return imgdata.Data, nil
			}

			if err = wm.RgbColourspace(); err != nil {
				logx.Error("ApplyWaterMark wm.RgbColourspace err=", err, " waterMarkSource=", waterMarkSource, " picId=", picId)
				return imgdata.Data, nil
			}

			if userInfo.WatermarkingType == 2 { // 水印居中
				err = wm.Embed(gravityCenter, width, height, 20, 20, rgbColor{0, 0, 0})
			} else {
				err = wm.Embed(gravitySouthEast, width, height, 20, 20, rgbColor{0, 0, 0})
			}
			if err != nil {
				logx.Error("ApplyWaterMark wm.Embed err=", err, " picId=", picId)
				return imgdata.Data, nil
			}

			if err := img.ApplyWatermark(wm, 1); err != nil {
				logx.Error("ApplyWaterMark img.ApplyWatermark err=", err, " picId=", picId)
				return imgdata.Data, nil
			}

			bufLast, buffCancel, err := img.Save(imgdata.Type, 100)
			if err != nil {
				logx.Error("ApplyWaterMark img.Save err=", err, " picId=", picId)
				return imgdata.Data, nil
			}

			return bufLast, buffCancel
		}
	}

	return imgdata.Data, nil
}
```

**Reply by lifang218c** [reply]
it sometimes report the error, when i call img.save function, is switch the picture type to call save functions.

**Reply by lifang218c** [reply]
by the way，i find that  color space。 i somestimes get this warining：

`GLib-GObject-WARNING **: 10:42:47.259: value "0" of type 'gint' is invalid or out of range for property 'colours' of type 'gint'`

**Reply by sethify** [reply]
I'm going to guess you use a byte slice created in GO, and create a vips image using this buffer. I'm going to guess that the byte slice is lost (goes out of scope) after the vips image is created. 

For example, calling a GO function that loads a file, and creates and returns vips image from the file will fail (but not always because of the garbage collector). The vips image will be valid but the byte slice will be freed, then depending on memory pressure that memory may be quickly reused. GO thinks you are done with the slice and releases it (GO cannot tell that libvips is using the pointer to the bytes in the byte slice). 

For us, the issue was that tests would just occasionally fail if we were doing parallel tests. This was because the memory used by the byte slice would get filled with other values. Libvips would warn, but not fail while doing operations on the image.

**Reply by lifang218c** [reply]
hi @sethify 
thanks for your reply~

img.save  must to be fail~. Because I saw the picture without watermark～ 

at the same time, when i call the : img.ApplyWatermark and img.embed  and so on before img.save,  i don't get error! it works fine, only the img.save report the error!!!   so  i don't  know how to handle it now ........

img.save is the last step to get the last picture data!!!!!!

**Comment by jcupitt** [comment]
Sorry, I don't know. As I said, the bug is probably in the Go binding you are using. I don't think this is a libvips issue.

I would ask the developer of your Go binding.

**Reply by lifang218c** [reply]
hi @jcupitt 

when i call the : img.ApplyWatermark and img.embed and so on before img.save, i don't get error! it works fine, only the img.save report the error!!!  not always error !!  about 10%, so i don't know how to handle it now

img.save is the last step to get the last picture data

**Reply by jcupitt** [reply]
I don't know, I didn't write any of the Go code you are using, but as Seth says, it sounds like a GC issue.

Please ask the developer of your Go binding.

**Reply by lifang218c** [reply]
hi, @jcupitt 

If it is a GC problem of go, why is this the only error reported in the last step. 
And if i request this picture alone,i can add watermark normally. the result is right......

**Reply by jcupitt** [reply]
libvips is a lazy library, so most processing is deferred until the final step. That's why you must keep a reference to the input memory until you do the final write, and that's why you are seeing GC problems.

I don't know what watermark code you are using, so I can't comment on that. I would ask the author.

**Comment by lifang218c** [comment]
i am  not sure, is this code cause this error :
```
runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	defer vipsCleanup()

func vipsCleanup() { C.vips_cleanup() }
```

**Reply by jcupitt** [reply]
Sorry, no idea, I don't know what `vips_cleanup()` is. I would ask the author.

**Comment by lifang218c** [comment]
defer runtime.KeepAlive(buf) 
i find some people saied to do this , but i don't try this.
i don't know if i do this maybe have some other effects.......
